From a6ad4074da3fa45ce3fbd084d44d228e1d5526c0 Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:43:09 +0800
Subject: [PATCH 1/4] GSI Android 17 compatibility:
 0001-Act-as-usb-device-when-there-is-no-hal-but-we-believ.patch

Change-Id: I8a17dc65b703c21eda4e02a6d9bfd3aa1ef61a79
---
 .../usb/UsbConnectionBroadcastReceiver.java                | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/com/android/settings/connecteddevice/usb/UsbConnectionBroadcastReceiver.java b/src/com/android/settings/connecteddevice/usb/UsbConnectionBroadcastReceiver.java
index 16ea602..262afbc 100644
--- a/src/com/android/settings/connecteddevice/usb/UsbConnectionBroadcastReceiver.java
+++ b/src/com/android/settings/connecteddevice/usb/UsbConnectionBroadcastReceiver.java
@@ -93,6 +93,13 @@ public class UsbConnectionBroadcastReceiver extends BroadcastReceiver implements
             mFunctions = functions;
             mDataRole = mUsbBackend.getDataRole();
             mPowerRole = mUsbBackend.getPowerRole();
+                //If we have no USB HAL, mDataRole is invalid
+                //But we can't be connected AND have none data_role, so it's safe.
+		//It would be better to fix UsbManager when no HAL is available, but that's more work
+                if(mDataRole == UsbPortStatus.DATA_ROLE_NONE &&
+                        intent.getExtras().getBoolean(UsbManager.USB_CONNECTED) &&
+                        !intent.getExtras().getBoolean(UsbManager.USB_HOST_CONNECTED))
+                    mDataRole = UsbPortStatus.DATA_ROLE_DEVICE;
         } else if (UsbManager.ACTION_USB_PORT_CHANGED.equals(intent.getAction())) {
             UsbPortStatus portStatus = intent.getExtras()
                     .getParcelable(UsbManager.EXTRA_PORT_STATUS);
